src/wallet/wallet2.cpp

```diff
@@ -1459,24 +1460,30 @@ void wallet2::update_pool_state()
   MDEBUG("update_pool_state done first loop");
 
   // remove pool txes to us that aren't in the pool anymore
-  std::unordered_map<crypto::hash, wallet2::payment_details>::iterator uit = m_unconfirmed_payments.begin();
-  while (uit != m_unconfirmed_payments.end())
+  // but only if we just refreshed, so that the tx can go in
+  // the in transfers list instead (or nowhere if it just
+  // disappeared without being mined)
+  if (refreshed)
   {
-    const crypto::hash &txid = uit->second.m_tx_hash;
-    bool found = false;
-    for (const auto &it2: res.tx_hashes)
+    std::unordered_map<crypto::hash, wallet2::payment_details>::iterator uit = m_unconfirmed_payments.begin();
+    while (uit != m_unconfirmed_payments.end())
```

Comment:
From what I can see at the end of `wallet2::process_new_transaction`, the key of the multimap `m_unconfirmed_payments` seems to be the payment ID , so I think it's a mistake to query it with txids in the `for` loop a few lines below.

Also I can see @vtnerd's point that there's not much point in having `m_unconfirmed_payments` as a multimap since the only place it gets referenced is in `wallet2::get_unconfirmed_payments` where its entire content is simply copied to `std::list<std::pair<key,value>>`. I agree with @moneromooo-monero, however, that the overhead is negligible.